### DISCUSSION TITLE
Trying to make multipolygon from type=boundary

### DISCUSSION THREAD

**Discussion by sn4k3ch4rm3r** [systemed/tilemaker#504] [Q&A] [answered]
I'm trying to get Hungarian national parks to show correctly. 
Example: https://www.openstreetmap.org/relation/6106836
If I don't do much processing it's missing some parts:
![image](https://github.com/systemed/tilemaker/assets/39193138/f2e7e4e3-383a-47b7-840d-0da410555d49)
Using the `relation_scan_function` I was able to get every part of it, and the outline appears to be correct, but turning it into a polygon, it appears to be inside out:
![image](https://github.com/systemed/tilemaker/assets/39193138/d8f34bfe-09e4-43fa-b206-5acc1dbbd281)

This is the process.lua I was able to put together based on the docs and the OMT process.lua
```lua
function relation_scan_function(relation)
	if relation:Find("type")=="boundary" and relation:Find("boundary")=="national_park" then
		relation:Accept()
	end
end

function way_function(way)
	local isBoundary = false
	local name = ""
	while true do
		local rel = way:NextRelation()
		if not rel then break end
		isBoundary = true
		if name == "" then
			name = way:FindInRelation("name")
		end
	end

	if isBoundary then
		way:Layer("park", true)
		way:Attribute("class", "national_park")
		way:Attribute("name", name)
	end
end
```

**Comment by systemed** [comment, marked as answer]
I doubt that'll work like that. tilemaker has special handling for multipolygon relations (i.e. `type=multipolygon`). It looks like boundary relations probably need to be processed that way, rather than with `relation_scan_function`.

I imagine you could get some way towards this by patching read_pbf.cpp:

(in PbfReader::ScanRelations)

````c++
	int typeKey = findStringPosition(pb, "type");
	int mpKey   = findStringPosition(pb, "multipolygon");
	int boundaryKey = findStringPosition(pb, "boundary");

	for (int j=0; j<pg.relations_size(); j++) {
		Relation pbfRelation = pg.relations(j);
		bool isMultiPolygon = (find(pbfRelation.keys().begin(), pbfRelation.keys().end(), typeKey) != pbfRelation.keys().end()) &&
		                      ((find(pbfRelation.vals().begin(), pbfRelation.vals().end(), mpKey ) != pbfRelation.vals().end()) ||
		                       (find(pbfRelation.vals().begin(), pbfRelation.vals().end(), boundaryKey) != pbfRelation.vals().end()));
````

(in PbfReader::ReadRelations)

````c++
		int typeKey = findStringPosition(pb, "type");
		int mpKey   = findStringPosition(pb, "multipolygon");
		int boundaryKey = findStringPosition(pb, "boundary");
		int innerKey= findStringPosition(pb, "inner");
		//int outerKey= findStringPosition(pb, "outer");
		if (typeKey >-1 && mpKey>-1) {
			for (int j=0; j<pg.relations_size(); j++) {
				Relation pbfRelation = pg.relations(j);
				bool isMultiPolygon = (find(pbfRelation.keys().begin(), pbfRelation.keys().end(), typeKey) != pbfRelation.keys().end()) &&
		                      ((find(pbfRelation.vals().begin(), pbfRelation.vals().end(), mpKey ) != pbfRelation.vals().end()) ||
		                       (find(pbfRelation.vals().begin(), pbfRelation.vals().end(), boundaryKey) != pbfRelation.vals().end()));
````

(I haven't tried this!)

**Reply by sn4k3ch4rm3r** [reply]
Thank you, it seems to have worked.

**Reply by systemed** [reply]
Excellent. I'll look at making this a permanent fix.

**Comment by BenOnTrack** [comment]
Hi, do you mind sharing your solution @sn4k3ch4rm3r ? 
I have the exact same issue with https://www.openstreetmap.org/relation/3659121#map=12/-36.9576/174.5459

If we should handle the "boundary" as multipolygon with the solution provided above, then how can I pick the ways to re-build this relation R3659121 and store them in a specific layer knowing those ways (in my example) have no tags?

thanks in advance for your feedback

**Reply by systemed** [reply]
You should find that https://github.com/systemed/tilemaker/pull/508 fixes this. ~You will need your `way_function` to look for `boundary=protected_area` and to act accordingly.~

**Reply by systemed** [reply]
I've reworked #508 a bit and you can now handle a multipolygon boundary by:

- accepting the boundary in `relation_scan_function`
- writing it to a layer in `relation_function` (not way_function)

tilemaker will pick up on `type=boundary` and the inner/outer roles, and automatically handle it as a multipolygon geometry.

There's an example of how to do this in the comment for #508.

**Reply by BenOnTrack** [reply]
It works great, I'm so excited !
Thanks so much @systemed 
Ben.
